pkg/controller/notification/mailgun_notification_delivery_service.go

```diff
@@ -2,17 +2,76 @@ package notification
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"github.com/mailgun/mailgun-go/v4"
+	"time"
 )
 
 type MailgunNotificationDeliveryService struct {
+	base        BaseNotificationDeliveryService
+	Mailgun     mailgun.Mailgun
+	SenderEmail string
 }
 
-// NewMailgunNotificationDeliveryService
-func NewMailgunNotificationDeliveryService() NotificationDeliveryService {
-	return &MailgunNotificationDeliveryService{}
+type MailgunConfiguration interface {
+	GetMailgunDomain() string
+	GetMailgunAPIKey() string
+	GetMailgunSenderEmail() string
 }
 
-func (s *MailgunNotificationDeliveryService) Send(ctx context.Context, notificationCtx NotificationContext, templateName string) error {
-	// TODO implement
-	return nil
+type MailgunOption interface {
+	// ApplyToMailgun applies this configuration to the given mailgun instance.
+	ApplyToMailgun(mailgun.Mailgun)
+}
+
+// NewMailgunNotificationDeliveryService creates a delivery service that uses the Mailgun API to deliver email notifications
+func NewMailgunNotificationDeliveryService(config MailgunConfiguration, templateLoader TemplateLoader,
+	opts ...MailgunOption) NotificationDeliveryService {
+
+	mg := mailgun.NewMailgun(config.GetMailgunDomain(), config.GetMailgunAPIKey())
+
+	for _, opt := range opts {
+		opt.ApplyToMailgun(mg)
+	}
+
+	svc := &MailgunNotificationDeliveryService{
```

Comment:
we can probably omit the svc var and just return &MailgunNotificationDeliveryService{ ... } here, what do you think?

Reply:
Good catch, fixed.